### PR TITLE
Add QuantumCircuit method to compute estimated fidelity

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds a new method to the quantum circuit (in both Python and C) to easily compute the estimated fidelity of a physical circuit for a given target. This uses the fidelity estimation that Qiskit's internal passes such as VF2Layout, VF2PostLayout, and TwoQubitPeepholeOptimization use for estimating fidelity of a circuit or subcircuit. This is a common operation when working with circuit compilation, especially when trying different pass manager constructions, or pass configurations to try and optimize the circuit for execution on a given backend. Having a quick metric for comparison of how the transpiler reasons about different compilations of the same circuit is useful for debugging compilation workflows.

While writing this function is fairly straightforward to do in Python with a simple for loop like:

```python
fidelity = 1.0
for inst in circuit.data:
    qargs = tuple(circuit.find_bits(x).index for x in inst.qubits)
    fidelity *= 1 - target[inst.name][qargs].error
```
(it's a bit more involved in C but fairly trivial too) given the general utility of this and avoiding Python iteration for large circuits writing this in Rust and exposing it to both C and Python for a consistent method is worth the addition.

The rust implementation of this new function is put in `qiskit_transpiler::target::estimate_fidelity` mostly because that was the lowest part of the internal dependency tree that it could be implemented. Typically this kind of function we would put in the `qiskit_accelerate` crate as a random accelerator, however because this function is being exposed to C too that was not a viable destination for it. Since we need both `CircuitData` and `Target` available to implement the function this made the most sense. The other alternative is probably to create a standalone module in the transpiler crate but this function is so small that didn't seem necessary.

### AI/LLM disclosure

- [X] I didn't use LLM tooling, or only used it privately.
- [ ] I used the following tool to help write this PR description:
- [ ] I used the following tool to generate or modify code:

### CODE REVIEW

File: crates/transpiler/src/target/mod.rs [hunk 1]
```diff
@@ -1678,10 +1679,69 @@ where
...
+            // SAFETY: The base type of Qubit and PhysicalQubit are both u32
+            // so they are memory compatible. The transmute call lets us treat
+            // Qubit qargs as PhysicalQubit qargs directly. This function as documented as
+            // assuming that the circuit is physical so the qubit index is a physical qubit on the
+            // target
+            let physical_qubits: &[PhysicalQubit] = unsafe { std::mem::transmute(qubits) };
+            target.get_error(inst.op.name(), physical_qubits)
+        })
+        .collect();
+    errors.map(|errors| 1. - fast_product(&errors))
```
Comment (jakelishman) [inline crates/transpiler/src/target/mod.rs]: Does the benefit of the SIMD actually outweigh the cost of the `Vec` collection for meaningfully sized circuits?

File: crates/transpiler/src/target/mod.rs [hunk 2]
```diff
@@ -1678,10 +1679,69 @@ where
...
+        .iter()
+        .filter(|inst| !inst.op.directive())
+        .map(|inst| {
+            let qubits = circuit.get_qargs(inst.qubits);
+            // SAFETY: The base type of Qubit and PhysicalQubit are both u32
+            // so they are memory compatible. The transmute call lets us treat
+            // Qubit qargs as PhysicalQubit qargs directly. This function as documented as
+            // assuming that the circuit is physical so the qubit index is a physical qubit on the
+            // target
+            let physical_qubits: &[PhysicalQubit] = unsafe { std::mem::transmute(qubits) };
```
Comment (jakelishman) [inline crates/transpiler/src/target/mod.rs]: I have a pretty old PR (https://github.com/Qiskit/qiskit/pull/13734) that makes `PhysicalQubit` wrap `Qubit` with `#[repr(transparent)]`, which I think is sufficient to let `bytemuck` do an unchecked safe cast.  We maybe want to revive that.

Minor nit: imo we shouldn't ever use type inference with such a dangerous function as `transmute`, because it moves some of the safety requirements (the types) to be non-local to the `unsafe` block.  I'd be slightly more comfortable if we could write
```rust
let physical_qubits = unsafe { std::mem::transmute::<&[Qubit], &[PhysicalQubit]>(qubits) };
```
so that both types are explicit and easily checkable with the SAFETY comment.

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -7745,6 +7747,41 @@ def estimate_duration(self, target, unit: str = "s") -> int | float:
             )
         return dur / prefix_dict[unit]
 
+    def estimate_fidelity(self, target) -> float | None:
+        """Estimate the fidelity of a physical circuit
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
        """Estimate the fidelity of a physical circuit.
```

File: crates/cext/src/circuit.rs [hunk 4]
```diff
@@ -2261,3 +2262,37 @@ pub unsafe extern "C" fn qk_circuit_copy_empty_like(
         .expect("Failed to copy the circuit.");
     Box::into_raw(Box::new(copied_circuit))
 }
+
+/// @ingroup QkCircuit
+/// Estimate the fidelity of a physical circuit
```
Comment (jakelishman) [inline crates/cext/src/circuit.rs]:
```suggestion
/// Estimate the fidelity of a physical circuit.
```

File: crates/cext/src/circuit.rs [hunk 5]
```diff
@@ -2261,3 +2262,37 @@ pub unsafe extern "C" fn qk_circuit_copy_empty_like(
...
+///
+/// This function will compute the product of the error rates for each
+/// gate in the circuit to estimate the fidelity of the circuit. This method is not
+///  intended to compute a realistic simulation of the fidelity of execution on real hardware. It is
+///  designed to provide an estimate of how the transpiler would work with the fidelity for various
+///  heuristics in it's operation. It is typically only useful for comparing different compilation
+///  outputs against each other to estimate which one would produce a better quality execution on
+///  hardware.
```
Comment (jakelishman) [inline crates/cext/src/circuit.rs]:
```suggestion
/// intended to compute a realistic simulation of the fidelity of execution on real hardware. It is
/// designed to provide an estimate of how the transpiler would work with the fidelity for various
/// heuristics in its operation. It is typically only useful for comparing different compilation
/// outputs against each other to estimate which one would produce a better quality execution on
/// hardware.
```

File: crates/cext/src/circuit.rs [hunk 6]
```diff
@@ -2261,3 +2262,37 @@ pub unsafe extern "C" fn qk_circuit_copy_empty_like(
...
+///
+/// @param circuit A pointer to the circuit to copy.
+/// @param target A pointer to the target that the circuit will be executed on. This is
+///     used to get the error rates for the instructions in the circuit. This function will
+///     return NaN if the circuit is not physical.
```
Comment (jakelishman) [inline crates/cext/src/circuit.rs]: Documentation of `circuit` doesn't look right.

File: crates/cext/src/circuit.rs [hunk 7]
```diff
@@ -2261,3 +2262,37 @@ pub unsafe extern "C" fn qk_circuit_copy_empty_like(
...
+///
+/// @return The computed fidelity of the circuit.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``circuit`` and ``target`` are not a valid, non-null pointer to a
+/// ``QkCircuit`` .
```
Comment (jakelishman) [inline crates/cext/src/circuit.rs]: Presumably this sentence should end with "or `QkTarget`, respectively"?

Fwiw, doxygen parses stuff as Markdown, so you only need one backtick, but it's no problem.

File: crates/cext/src/circuit.rs [hunk 8]
```diff
@@ -2261,3 +2262,37 @@ pub unsafe extern "C" fn qk_circuit_copy_empty_like(
...
+///  heuristics in it's operation. It is typically only useful for comparing different compilation
+///  outputs against each other to estimate which one would produce a better quality execution on
+///  hardware.
+///
+/// @param circuit A pointer to the circuit to copy.
+/// @param target A pointer to the target that the circuit will be executed on. This is
+///     used to get the error rates for the instructions in the circuit. This function will
+///     return NaN if the circuit is not physical.
+///
+/// @return The computed fidelity of the circuit.
```
Comment (jakelishman) [inline crates/cext/src/circuit.rs]: We're missing the explanation of a `nan` value.

File: crates/transpiler/src/target/mod.rs [hunk 9]
```diff
@@ -1678,10 +1679,69 @@ where
...
+            target.get_error(inst.op.name(), physical_qubits)
+        })
+        .collect();
```
Comment (jakelishman) [inline crates/transpiler/src/target/mod.rs]: `Target::get_error` returns `None` if _either_ the operation can't be found, or if the operation can be found but has no configured error.  The transpiler passes you named treat the first case as "missing" and the second case as "no error", but this logic is treating both cases as "missing" and returning `None`.

Can you add a test for a `Target` with mixed errors: some fully calibrated and some with `None` properties?

File: qiskit/circuit/quantumcircuit.py [hunk 10]
```diff
@@ -7745,6 +7747,41 @@ def estimate_duration(self, target, unit: str = "s") -> int | float:
...
+
+        This function will compute the product of the error rates for each
+        gate in the circuit to estimate the fidelity of the circuit:
+
+        ..math:: 1 - \\prod ^g error(g)
+          :label: error product
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
        ..math::
          :label: estimated circuit fidelity
          
          \\prod_{g \\in \\text{gates}} \\bigl(1 - \\operatorname{error}(g)\\bigr)
```

I started off listing like "this is not a) the correct formula, b) ..." but then I got to four separate incorrect elements...

File: qiskit/circuit/quantumcircuit.py [hunk 11]
```diff
@@ -7745,6 +7747,41 @@ def estimate_duration(self, target, unit: str = "s") -> int | float:
...
+
+        where $error(g)$ is the error rate in the target for the instruction $g$ from the circuit
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
        where :math:`\\operatorname{error}(g)` is the error rate in the target for the instruction :math:`g` from the circuit
```

(or `\text` if we don't support `\operatorname`)

File: qiskit/circuit/quantumcircuit.py [hunk 12]
```diff
@@ -7745,6 +7747,41 @@ def estimate_duration(self, target, unit: str = "s") -> int | float:
...
+        in the target. If the circuit is not physical, meaning any instruction in the circuit (as
+        in operation and qargs) is not found in the target, this will return ``None``. This method is not
+        intended to compute a realistic simulation of the fidelity of execution on real hardware. It is
+        designed to provide an estimate of how the transpiler would work with the fidelity for various
+        heuristics in it's operation. It is typically only useful for comparing different compilation
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
        heuristics in its operation. It is typically only useful for comparing different compilation
```

File: qiskit/circuit/quantumcircuit.py [hunk 13]
```diff
@@ -7745,6 +7747,41 @@ def estimate_duration(self, target, unit: str = "s") -> int | float:
...
+        outputs against each other to estimate which one would produce a better quality execution on
+        hardware.
+
+        Args:
+            target (Target): The :class:`.Target` instance that the circuit will be executed on and used to
+                get the error rates for the instructions in the circuit.
+
+        Returns:
+            The estimated fidelity of executing the circuit on the given target. If any
+            instruction in the circuit is not present in target
```
Comment (jakelishman) [inline qiskit/circuit/quantumcircuit.py:7776]: Sentence trails off

File: test/c/test_circuit.c [hunk 14]
```diff
@@ -1435,6 +1435,50 @@ static int test_pbc_instructions(void) {
...
+    qk_circuit_gate(qc, QkGate_X, (uint32_t[]){0}, NULL);
+    qk_circuit_gate(qc, QkGate_X, (uint32_t[]){0}, NULL);
+    qk_circuit_gate(qc, QkGate_X, (uint32_t[]){0}, NULL);
+    QkTarget *target = qk_target_new(1);
+    QkTargetEntry *entry = qk_target_entry_new(QkGate_X);
+    qk_target_entry_add_property(entry, (uint32_t[]){0}, 1, 0.5, 0.5);
+    qk_target_add_instruction(target, entry);
+    double fidelity = qk_circuit_estimate_fidelity(qc, target);
+    if (fidelity != (1. - 0.5 * 0.5 * 0.5)) {
+        printf("Expected 0.875 fidelity got %f instead\n", fidelity);
```
Comment (jakelishman) [inline test/c/test_circuit.c]: This is a weird mix of hardcoded and non-hardcoded values. (Also the expected value is incorrect - if you add _more_ x gates, you'll currently get a "better" result!)

File: crates/transpiler/src/target/mod.rs (hunk 1 above)
Comment (jakelishman) [inline crates/transpiler/src/target/mod.rs]: This isn't a correct calculation of the fidelity.  $1 - \prod_i e_i$ isn't the same as $\prod_i (1 - e_i)$, which is what you mean.

Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: I'll remove this, the answer is that apparently there is no benefit to doing this:

simd_times

This was running on my Threadripper 3970x which has avx2. It might be different on other hardware though.

When I did a similar analysis for the sum of floats in #7702 the benefits were definitely there for smaller sizes and at larger sizes we used multithreading (and pulp was slightly faster when tested #11916).

File: crates/transpiler/src/target/mod.rs (hunk 9 above)
Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: Fixed and a test added in: https://github.com/Qiskit/qiskit/pull/16292/commits/4a5e386464759cbee02b6527d4ad22986e940e2b 

I actually think we might have a broader bug in the transpiler around this since `Target::get_error()` doesn't account for global non-ideal gates. So we are probably ignoring the error rates of gates added with properties like `{None: InstructionProperties(error=0.99999)}`. But that's something to fix in a separate PR.

File: crates/transpiler/src/target/mod.rs (hunk 2 above)
Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: Yeah we should revive that, this was the best way I could come up with what we have right now though. I updated the typing in https://github.com/Qiskit/qiskit/pull/16292/commits/4c21c4a523b32cafa351dad488d9cccad95e4d8f to be explicit instead of relying on inference.

File: crates/transpiler/src/target/mod.rs [hunk 15]
```diff
@@ -1678,10 +1678,86 @@ where
...
+        .map(|inst| {
+            let qubits = circuit.get_qargs(inst.qubits);
+            let gate_name = inst.op.name();
+            // SAFETY: The base type of Qubit and PhysicalQubit are both u32
+            // so they are memory compatible. The transmute call lets us treat
+            // Qubit qargs as PhysicalQubit qargs directly. This function as documented as
+            // assuming that the circuit is physical so the qubit index is a physical qubit on the
+            // target
+            let physical_qubits: &[PhysicalQubit] =
+                unsafe { std::mem::transmute::<&[Qubit], &[PhysicalQubit]>(qubits) };
```
Comment (jakelishman) [inline crates/transpiler/src/target/mod.rs]: As it transpires, I did already write this, regardless of the other PR: it's called `PhysicalQubit::lift_slice`.

**Review by jakelishman — COMMENTED** [review]
(not a full re-review, just commenting something before I forget it)

File: crates/transpiler/src/target/mod.rs (hunk 15 above)
Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: I switched to it in: https://github.com/Qiskit/qiskit/pull/16292/commits/7ba40f545483a8f69338c57b48ce7f3f76442fe1

File: qiskit/circuit/quantumcircuit.py (hunk 13 above)
Comment (mtreinish) [reply qiskit/circuit/quantumcircuit.py:7776]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/646a7f68dff50e1682ec6dce6615445013ff506a

File: crates/cext/src/circuit.rs (hunk 8 above)
Comment (mtreinish) [reply crates/cext/src/circuit.rs]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/646a7f68dff50e1682ec6dce6615445013ff506a for some reason I put it on the `target` argument description originally :shrug:

File: crates/cext/src/circuit.rs (hunk 7 above)
Comment (mtreinish) [reply crates/cext/src/circuit.rs]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/646a7f68dff50e1682ec6dce6615445013ff506a

File: crates/cext/src/circuit.rs (hunk 6 above)
Comment (mtreinish) [reply crates/cext/src/circuit.rs]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/646a7f68dff50e1682ec6dce6615445013ff506a

File: crates/cext/src/circuit.rs (hunk 4 above)
Comment (mtreinish) [reply crates/cext/src/circuit.rs]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/646a7f68dff50e1682ec6dce6615445013ff506a

File: crates/transpiler/src/target/mod.rs [hunk 16]
```diff
@@ -1677,10 +1677,85 @@ where
     obj.eq(other.into_bound_py_any(py)?)
 }
 
+/// Estimate the fidelity of a circuit by taking the product of the error rates
+///
+/// It is assumed that calling this function the circuit is already physical (already transpiled)
+/// such that the qubits for each gate correspond to the physical qubits in the target.
```
Comment (alexanderivrii) [inline crates/transpiler/src/target/mod.rs]: It seems that "when" is missing.

```suggestion
/// It is assumed that, when this function is called, the circuit is already physical (already transpiled)
/// such that the qubits for each gate correspond to the physical qubits in the target.
```

File: crates/transpiler/src/target/mod.rs [hunk 17]
```diff
@@ -1677,10 +1677,85 @@ where
...
+///     The estimated fidelity. If any operation can't be found in the target it will return None.
+#[pyfunction(name = "estimate_fidelity")]
+pub fn py_estimate_fidelity(circuit: &PyCircuitData, target: &Target) -> Option<f64> {
+    estimate_fidelity(circuit, target)
+}
+
+/// Estimate the fidelity of a circuit by taking the product of the error rates
+///
+/// It is assumed that calling this function the circuit is already physical (already transpiled)
+/// such that the qubits for each gate correspond to the physical qubits in the target.
```
Comment (alexanderivrii) [inline crates/transpiler/src/target/mod.rs]: same here

File: crates/transpiler/src/target/mod.rs [hunk 18]
```diff
@@ -1677,10 +1677,85 @@ where
...
+pub fn estimate_fidelity(circuit: &CircuitData, target: &Target) -> Option<f64> {
+    circuit
+        .data()
+        .iter()
+        .filter(|inst| !inst.op.directive())
+        .map(|inst| {
+            let qubits = circuit.get_qargs(inst.qubits);
+            let gate_name = inst.op.name();
+            // SAFETY: The base type of Qubit and PhysicalQubit are both u32
+            // so they are memory compatible. The transmute call lets us treat
```
Comment (alexanderivrii) [inline crates/transpiler/src/target/mod.rs]: do you want to mention the `transmute` here now that this uses `lift_slice`?

File: test/python/circuit/test_circuit_operations.py [hunk 19]
```diff
@@ -1786,6 +1786,68 @@ def test_ensure_physical_cannot_shrink_circuit(self):
...
+            qc.x(0)
+        with self.assertRaises(CircuitError):
+            qc.estimate_fidelity(target)
+
+    def test_estimate_fidelity_with_ideal_gates(self):
+        target = Target(num_qubits=1)
+        target.add_instruction(XGate(), {(0,): InstructionProperties(error=0.25)})
+        target.add_instruction(YGate(), {(0,): InstructionProperties(error=None)})
+        target.add_instruction(ZGate())
+        target.add_instruction(SXGate(), {None: InstructionProperties(error=0.1)})
```
Comment (alexanderivrii) [inline test/python/circuit/test_circuit_operations.py:1841]: Nice! This test checks both ideal gates (Z) and global gates (SX).

**Review by alexanderivrii — COMMENTED** [review]
Thanks Matthew, I am already using this PR for various experiments.

LGTM, modulo a minor suggestion about the docstring. 

There is also something wrong with the python calculation in the PR summary, it should be more like
```python
fidelity = 1.0
for inst in circuit.data:
    qargs = tuple(circuit.find_bits(x).index for x in inst.qubits)
    fidelity *= 1 - target[inst.name][qargs].error
``` 
but the rust code is completely correct and the tests cover all possible branches in the computation: gate appears in the instruction properties, gate is global, gate is ideal, gate is unsupported.

File: crates/transpiler/src/target/mod.rs (hunk 18 above)
Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: Removed in: https://github.com/Qiskit/qiskit/pull/16292/commits/fdc5a4787c88410b4ed1805073910386f145c842 the comment was leftover from an earlier iteration where I had an unsafe block using transmute to do the slice casting without a copy.

File: crates/transpiler/src/target/mod.rs (hunk 17 above)
Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/b6b0bc36d608ffdb2ef9e7a4e49bd09573ceda9c

File: crates/transpiler/src/target/mod.rs (hunk 16 above)
Comment (mtreinish) [reply crates/transpiler/src/target/mod.rs]: Fixed in: https://github.com/Qiskit/qiskit/pull/16292/commits/b6b0bc36d608ffdb2ef9e7a4e49bd09573ceda9c

**Comment by mtreinish** [comment]
> There is also something wrong with the python calculation in the PR summary, it should be more like

I've fixed this in the PR summary. When I originally opened this PR I mistakenly used the wrong formula without thinking about it too much and then copied it in every place including the commit message. I fixed that after Jake pointed out my mistake in his initial review, but I forgot to update the PR summary.

**Review by alexanderivrii — APPROVED** [review]
Thanks!
